nvflare/apis/utils/task_utils.py

```diff
@@ -0,0 +1,49 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nvflare.apis.fl_constant import FilterKey, FLContextKey
+
+
+def _apply_filters(filter_list, filter_error, filter_data, fl_ctx):
+    if filter_list:
+        for f in filter_list:
+            filter_data = f.process(filter_data, fl_ctx)
+    return filter_error, filter_data
+
+
+def apply_data_filters(data_filters, task_data, fl_ctx, task_name, direction):
```

Comment:
The app_data_filters and apply_result_filters are very similar: only difference is which filter to use from the scope object. Consider refactoring to use a common func.

Reply:
extract the common function as much as we can.

Reply:
If we want we could do this:


```
def _apply_filters(filters_name, result_filters, result, fl_ctx, task_name, direction):
    filter_list = []
    scope_object = fl_ctx.get_prop(FLContextKey.SCOPE_OBJECT)
    if scope_object:
        filters = getattr(scope_object, filters_name)
        if filters:
            filter_list.extend(filters.get(direction, []))
    task_filter_list = config_filters.get(task_name + FilterKey.DELIMITER + direction)
    if task_filter_list:
        filter_list.extend(task_filter_list)

    if filter_list:
        for f in filter_list:
            filter_data = f.process(filter_data, fl_ctx)
    return filter_data

def apply_data_filters(result_filters, result, fl_ctx, task_name, direction):
    return _apply_filters("task_data_filters", result_filters, result, fl_ctx, task_name, direction)

def apply_result_filters(result_filters, result, fl_ctx, task_name, direction):
    return _apply_filters("task_result_filters", result_filters, result, fl_ctx, task_name, direction)

```

Reply:
further extracted the common functions.